Documentation for the Democracy Module

ltfschoen opened the pull request:
Relates to https://github.com/paritytech/substrate-developer-hub/issues/28

Requires update for consistency with https://github.com/paritytech/substrate/pull/2172

joepetrowski commented:
A lot of code exceeded our style guide length limit. Fixed in most recent commit, but can be rolled back if we want this to be strictly comments.

ltfschoen commented:
@joepetrowski the indentation didn't look right in the code snippet when i ran ` cargo doc --package srml-democracy --open`. it was ok on lines where indentation had been done using spaces, but not when indentation had been done using tabs, so i just removed the tabs and replaced them with spaces

DemiMarie-temp commented on srml/democracy/src/vote_threshold.rs:
```diff
@@ -22,13 +22,13 @@ use parity_codec::{Encode, Decode};
 	/// A supermajority of approvals is needed to pass this vote.
 	SuperMajorityApprove,
-	/// A supermajority of rejects is needed to fail this vote.
+	/// A supermajority of rejections is needed to fail this vote.
```
How large a supermajority is needed?  This should be documented.

joepetrowski replied:
See update

ltfschoen commented:
i've received feedback from rpmeier suggesting that we providing information about what origin the proposed call is executed with

ltfschoen commented:
@shawntabrizi why are you force pushing and deleting all the commit history from other contributors? why not just close this PR and create a fresh branch of your own?

shawntabrizi commented:
@ltfschoen I am not intending to discredit any work here, but these docs have been stale and were originally opened to merge to the master branch.

Since the SRML Democracy module has changed quite a bit in master, we re-targeted this branch to merge to v1.0.

However, at certain points this branch was re-based or merged with the master branch which means that when we changed the base branch to v1.0, we were getting changes on the order of 400+ files changed.

I isolated the changes and fixed up the PR. There could probably be a better way to do this, but end of the day, I just want to get these docs merged in.

ltfschoen commented:
Fair enough,  thanks for cleaning it up!

ltfschoen commented on srml/democracy/src/lib.rs:
```diff
@@ -14,7 +14,143 @@
+//! - **Maturity:** A mature referendum is one that expires at the current block.
+//! - **Vote tallying, passing, and execution:** When a block is finalized, the Democracy module will search
+//!  for mature referenda and tally their votes, then remove them from the Table of Referenda. If a referendum's
+//!  vote tally meets its vote threshold, then it will be passed and executed. Last, increment `NextTally`
```
```suggestion
//!  vote tally meets its vote threshold, then it will be passed and executed. Lastly, `NextTally` is incremented
```

ltfschoen commented on srml/democracy/src/lib.rs:
```diff
@@ -14,7 +14,143 @@
+//!
+//! - `propose` - Propose a sensitive action to be taken.
+//! - `second` - Second (sponsor) a proposal.
+//! - `vote` - Vote in a referendum.
```
```suggestion
//! - `vote` - Vote on a referendum.
```

ltfschoen commented on srml/democracy/src/lib.rs:
```diff
@@ -14,7 +14,143 @@
+//! - `propose` - Propose a sensitive action to be taken.
+//! - `second` - Second (sponsor) a proposal.
+//! - `vote` - Vote in a referendum.
+//! - `proxy_vote` - Vote in a referendum on behalf of a stash.
```
```suggestion
//! - `proxy_vote` - Vote on a referendum on behalf of a stash account.
```

ltfschoen commented on srml/democracy/src/lib.rs:
```diff
@@ -14,7 +14,143 @@
+//! - `cancel_referendum` - Remove a referendum.
+//! - `cancel_queued` - Cancel a proposal queued for enactment.
+//! - `on_finalize` - Called when a block is finalized.
+//! - `set_proxy` - Specify a proxy. Called by the stash.
```
```suggestion
//! - `set_proxy` - Specify a proxy account. Called by the stash account.
```

ltfschoen replied:
nitpick: if added consider updating implementation too https://github.com/paritytech/substrate/pull/2286/files#diff-cbe2b253de4fb4f87ccc41b9075568d6R304

ltfschoen commented on srml/democracy/src/lib.rs:
```diff
@@ -14,7 +14,143 @@
+//! - `cancel_queued` - Cancel a proposal queued for enactment.
+//! - `on_finalize` - Called when a block is finalized.
+//! - `set_proxy` - Specify a proxy. Called by the stash.
+//! - `resign_proxy` - Clear the proxy. Called by the proxy.
```
```suggestion
//! - `resign_proxy` - Clear the proxy account. Called by the proxy account.
```

ltfschoen commented on srml/democracy/src/lib.rs:
```diff
@@ -14,7 +14,143 @@
+//! - `on_finalize` - Called when a block is finalized.
+//! - `set_proxy` - Specify a proxy. Called by the stash.
+//! - `resign_proxy` - Clear the proxy. Called by the proxy.
+//! - `remove_proxy` - Clear the proxy. Called by the stash.
```
```suggestion
//! - `remove_proxy` - Clear the proxy account. Called by the stash account.
```

ltfschoen commented on srml/democracy/src/lib.rs:
```diff
@@ -14,7 +14,143 @@
+//! ### Public Functions
+//!
+//! - `locked_for` - Get the balance locked in support of a proposal.
+//! - `is_active_referendum` - Return true if given index corresponds to an on-going referendum.
```
```suggestion
//! - `is_active_referendum` - Return true if the given referendum index corresponds to an on-going referendum.
```

ltfschoen commented on srml/democracy/src/lib.rs:
```diff
@@ -118,22 +254,26 @@ decl_module! {
 		}
 
-		/// Vote in a referendum on behalf of a stash. If `vote.is_aye()`, the vote is to enact the proposal;
+		/// Vote in a referendum on behalf of a stash. If `vote.is_aye`, the vote is to enact the proposal;
```
```suggestion
		/// Vote on a referendum using a proxy account on behalf of a stash account.. If `vote.is_aye`, the vote is to enact the proposal;
```

ltfschoen commented on srml/democracy/src/lib.rs:
```diff
@@ -118,22 +254,26 @@ decl_module! {
 		}
 
-		/// Vote in a referendum. If `vote.is_aye()`, the vote is to enact the proposal;
+		/// Vote in a referendum. If `vote.is_aye`, the vote is to enact the proposal;
```
```suggestion
		/// Vote on a referendum. If `vote.is_aye`, the vote is to enact the proposal;
```

ltfschoen commented on srml/democracy/src/lib.rs:
```diff
@@ -14,7 +14,143 @@
+//! decl_module! {
+//! 	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+//!
+//! 		fn proxy_set_approvals(origin, votes: Vec<bool>, index: VoteIndex) -> Result {
```
nitpick: is this function indented too much?

shawntabrizi replied:
I think there are spacing issues in GitHub which do not appear when you open it in Visual Studio Code.

ltfschoen commented on srml/democracy/src/lib.rs:
```diff
@@ -14,7 +14,143 @@
+//!
+//! decl_module! {
+//! 	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+//!
```
```suggestion
//! 		/// Specify approval votes for a given referendum index. Called by a proxy account on behalf of a stash account.
```

ltfschoen commented on srml/democracy/src/lib.rs:
```diff
@@ -14,7 +14,143 @@
+//! decl_module! {
+//! 	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+//!
+//! 		fn proxy_set_approvals(origin, votes: Vec<bool>, index: VoteIndex) -> Result {
```
```suggestion
//! 		fn set_approvals(origin, votes: Vec<bool>, index: ReferendumIndex) -> Result {
```

ltfschoen commented on srml/democracy/src/lib.rs:
```diff
@@ -153,6 +293,8 @@ decl_module! {
 			<DispatchQueue<T>>::mutate(when, |items| if items.len() > which { items[which] = None });
 		}
 
+		/// Called when a block is finalized. Will launch a new referendum if it is time, finish
```
```suggestion
		/// Called when a block is finalized. Launch a new referendum if it is time to, finish
```

ltfschoen commented on srml/democracy/src/lib.rs:
```diff
@@ -239,7 +398,8 @@ decl_storage! {
 		pub PublicDelay get(public_delay) config(): T::BlockNumber;
-		/// The maximum number of additional lock periods a voter may offer to strengthen their vote. Multiples of `PublicDelay`.
+		/// The maximum number of additional lock periods a voter may offer to strengthen
+		/// their vote. Multiples of `PublicDelay`.
```
```suggestion
		/// their vote. `PublicDelay` is a multiple of `LockPeriods`.
```

ltfschoen replied:
i don't think ""Multiples of `Public`" delay makes sense, since:
* `PublicDelay` is measured using `BlockNumber`, which is type `u64`.
* `LockPeriods` is type `i8`.

ltfschoen commented on srml/democracy/src/lib.rs:
```diff
@@ -285,9 +445,9 @@ decl_event!(
+	// Exposed immutables.
 
-	/// Get the amount locked in support of `proposal`; `None` if proposal isn't a valid proposal
+	/// Get the balance locked in support of `proposal`; `None` if proposal isn't a valid proposal
```
```suggestion
	/// Get the balance locked in support of `proposal`; `None` if proposal isn't a valid proposal.
```

ltfschoen replied:
```suggestion
	/// Get the balance locked in support of `proposal`; `None` if proposal isn't a valid proposal index.
```

ltfschoen commented on srml/democracy/src/lib.rs:
```diff
@@ -285,9 +445,9 @@ decl_event!(
 
-	/// Get the amount locked in support of `proposal`; `None` if proposal isn't a valid proposal
+	/// Get the balance locked in support of `proposal`; `None` if proposal isn't a valid proposal
 	/// index.
```
```suggestion
```

ltfschoen replied:
i think this can fit on one line

ltfschoen commented on srml/democracy/src/lib.rs:
```diff
@@ -431,29 +615,44 @@ impl<T: Trait> Module<T> {
 		Self::deposit_event(RawEvent::Executed(index, ok));
 	}
 
+	/// Finds the proposal that will be the next referendum, unreserves the currency of
```
```suggestion
	/// Find the proposal that will be the next referendum, unreserves the currency of
```

ltfschoen replied:
nitpick: just for consistency, since we're saying `Tally` not `Tallying` in other doc comments

ltfschoen commented on srml/democracy/src/lib.rs:
```diff
@@ -700,7 +906,7 @@ mod tests {
 			assert_noop!(Democracy::remove_proxy(Origin::signed(2), 10), "wrong proxy");
 
-			// 1 fires his proxy:
+			// 1 fires their proxy:
```
```suggestion
			// stash account 1 fires their proxy:
```

ltfschoen replied:
nitpick: i also think `their` should be replaced with `its`

ltfschoen commented on srml/democracy/src/vote_threshold.rs:
```diff
@@ -22,13 +22,15 @@ use parity_codec::{Encode, Decode};
 use rstd::ops::{Add, Mul, Div, Rem};
 
-/// A means of determining if a vote is past pass threshold.
+/// A means of determining if a vote is past its pass threshold.
```
```suggestion
/// A means of determining if a vote is above the required threshold to pass.
```

ltfschoen commented on srml/democracy/src/vote_threshold.rs:
```diff
@@ -69,14 +73,38 @@ fn compare_rationals<T: Zero + Mul<T, Output = T> + Div<T, Output = T> + Rem<T,
+	/// - a total electorate size of `electorate`, and
+	/// - the number of `voters` who voted,
+	///
+	/// this function will return true if the overall outcome is in favor of approval.
```
```suggestion
	/// Return true if the overall outcome is in favor of approval.
```

ltfschoen commented on srml/democracy/src/vote_threshold.rs:
```diff
@@ -69,14 +73,38 @@ fn compare_rationals<T: Zero + Mul<T, Output = T> + Div<T, Output = T> + Rem<T,
+	///
+	/// - `approve` votes for a proposal,
+	/// - `against` votes against a proposal,
+	/// - a total electorate size of `electorate`, and
```
```suggestion
	/// - A total electorate size of `electorate`; and
```

ltfschoen commented on srml/democracy/src/vote_threshold.rs:
```diff
@@ -69,14 +73,38 @@ fn compare_rationals<T: Zero + Mul<T, Output = T> + Div<T, Output = T> + Rem<T,
+{
+	/// Given:
+	///
+	/// - `approve` votes for a proposal,
```
```suggestion
	/// - Votes that `approve` a proposal.
```

ltfschoen commented on srml/democracy/src/vote_threshold.rs:
```diff
@@ -69,14 +73,38 @@ fn compare_rationals<T: Zero + Mul<T, Output = T> + Div<T, Output = T> + Rem<T,
+	/// Given:
+	///
+	/// - `approve` votes for a proposal,
+	/// - `against` votes against a proposal,
```
```suggestion
	/// - Vote that are `against` a proposal.
```

ltfschoen commented on srml/democracy/src/vote_threshold.rs:
```diff
@@ -69,14 +73,38 @@ fn compare_rationals<T: Zero + Mul<T, Output = T> + Div<T, Output = T> + Rem<T,
+	/// - `approve` votes for a proposal,
+	/// - `against` votes against a proposal,
+	/// - a total electorate size of `electorate`, and
+	/// - the number of `voters` who voted,
```
```suggestion
	/// - Number of `voters` who voted.
```

ltfschoen commented on srml/democracy/src/vote_threshold.rs:
```diff
@@ -42,7 +44,9 @@ pub trait Approved<Balance> {
 /// Return `true` iff `n1 / d1 < n2 / d2`. `d1` and `d2` may not be zero.
-fn compare_rationals<T: Zero + Mul<T, Output = T> + Div<T, Output = T> + Rem<T, Output = T> + Ord + Copy>(mut n1: T, mut d1: T, mut n2: T, mut d2: T) -> bool {
+fn compare_rationals<T>(mut n1: T, mut d1: T, mut n2: T, mut d2: T) -> bool
+	where T: Zero + Mul<T, Output = T> + Div<T, Output = T> + Rem<T, Output = T> + Ord + Copy
```
suggest this be included in a separate PR since it changes the implementation (instead of mixed into a documentation-specific PR)

shawntabrizi replied:
It does not change anything really, it is just a reformat of the code

ltfschoen commented on srml/democracy/src/vote_threshold.rs:
```diff
@@ -69,14 +73,38 @@ fn compare_rationals<T: Zero + Mul<T, Output = T> + Div<T, Output = T> + Rem<T,
-	/// `electorate` of whom `voters` voted (`electorate - voters` are abstainers) then returns true if the
-	/// overall outcome is in favor of approval.
+impl<Balance> Approved<Balance> for VoteThreshold
+	where Balance: IntegerSquareRoot
```
suggest this be included in a separate PR since it changes the implementation (instead of mixed into a documentation-specific PR)

ltfschoen commented:
With regard to the actual implementation, I came up with the following questions:

1) In the implementation of `set_proxy` https://github.com/paritytech/substrate/pull/2286/files#diff-cbe2b253de4fb4f87ccc41b9075568d6R305, how are we checking that it is actually being called by a **stash account** as expected (rather than being called by a **proxy account** for instance)? do we need an additional assertion?

2) In the implementation of `resign_proxy` https://github.com/paritytech/substrate/pull/2286/files#diff-cbe2b253de4fb4f87ccc41b9075568d6R312, how are we checking that it is actually being called by a **proxy account** as expected (rather than being called by a **stash account** for instance)? do we need an additional assertion?

3) In the implementation of `remove_proxy` https://github.com/paritytech/substrate/pull/2286/files#diff-cbe2b253de4fb4f87ccc41b9075568d6L176, how are we checking that it is actually being called by a **stash account** as expected (rather than being called by a **proxy account** for instance)? do we need an additional assertion?

shawntabrizi commented on srml/democracy/src/lib.rs:
```diff
@@ -14,7 +14,143 @@
+//!
+//! ## Genesis Config
+//!
+//! The Democracy module depends on the [`GenesisConfig`](./struct.GenesisConfig.html).
```
Sounds good

shawntabrizi commented on srml/democracy/src/lib.rs:
```diff
@@ -14,7 +14,143 @@
+//!  by signing and submitting an extrinsic. It is allocated the next referendum index, which is mapped to its
+//!  corresponding voting period expiry block, the proposal it relates to, and the given voting threshold.
+//! - **Cancellation:** Remove all information about a referendum.
+//! - **Validity:** The new referendum must not have a voting period that ends before any existing referenda.
```
No, the current line makes sense based on the code.

		if ref_index > 0 && Self::referendum_info(ref_index - 1).map(|i| i.end > end).unwrap_or(false) {
			Err("Cannot inject a referendum that ends earlier than preceeding referendum")?
		}
